File: modules/vescape-core/ios/accessory/AccessoryFixtures.swift
import Foundation

/// The shared Accessory Protocol corpus, located relative to this file the way the Refloat schema
/// fixtures are. The same files drive the Kotlin peer and the ESP32 firmware's native tests, so a
/// contract that drifts on one side fails on all three.
///
/// @parity /modules/vescape-core/android/src/test/java/expo/modules/vescapecore/accessory/AccessoryFixtures.kt
enum AccessoryFixtures {
  static func load(_ name: String) throws -> [String: Any] {
    let root = URL(fileURLWithPath: #filePath)
      .deletingLastPathComponent()  // accessory
      .deletingLastPathComponent()  // ios
      .deletingLastPathComponent()  // vescape-core
      .deletingLastPathComponent()  // modules
      .deletingLastPathComponent()  // repo root
    let data = try Data(
      contentsOf: root.appendingPathComponent("shared/fixtures/accessory-protocol/\(name)")
    )
    guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      throw NSError(
        domain: "AccessoryFixtures", code: 1,
        userInfo: [NSLocalizedDescriptionKey: "\(name) is not a JSON object"]
      )
    }
    return object
  }

  static func hexToBytes(_ hex: String) -> [UInt8] {
    var bytes: [UInt8] = []
    bytes.reserveCapacity(hex.count / 2)
    var index = hex.startIndex
    while index < hex.endIndex {
      let next = hex.index(index, offsetBy: 2)
      bytes.append(UInt8(hex[index..<next], radix: 16) ?? 0)
      index = next
    }
    return bytes
  }
}

File: modules/vescape-core/android/src/test/java/expo/modules/vescapecore/accessory/AccessoryFixtures.kt
package expo.modules.vescapecore.accessory

import java.io.File
import org.json.JSONObject

/**
 * The shared Accessory Protocol corpus, read straight off the repo tree the way the Refloat schema
 * fixtures are. The same files drive the Swift peer and the ESP32 firmware's native tests, so a
 * contract that drifts on one side fails on all three.
 *
 * @parity /modules/vescape-core/ios/accessory/AccessoryFixtures.swift
 */
internal object AccessoryFixtures {
    private const val DIR = "shared/fixtures/accessory-protocol"

    fun load(name: String): JSONObject {
        val file = File(repoRoot(), "$DIR/$name")
        require(file.isFile) { "missing accessory fixture $name" }
        return JSONObject(file.readText())
    }

    private fun repoRoot(): File {
        var dir: File? = File(System.getProperty("user.dir")!!).absoluteFile
        while (dir != null && !File(dir, DIR).isDirectory) dir = dir.parentFile
        return requireNotNull(dir) { "$DIR not found above ${System.getProperty("user.dir")}" }
    }

    fun hexToBytes(hex: String): ByteArray {
        require(hex.length % 2 == 0) { "odd-length hex: $hex" }
        return ByteArray(hex.length / 2) {
            ((hex[it * 2].digitToInt(16) shl 4) or hex[it * 2 + 1].digitToInt(16)).toByte()
        }
    }
}
